climada/entity/exposures/litpop/nightlight.py

```diff
@@ -222,53 +224,121 @@ def download_nl_files(req_files=np.ones(len(BM_FILENAMES),),
-
-    Returns:
-        nightlight (sparse.csr_matrix), coord_nl (np.array)
+def load_nasa_nl_shape_single_tile(geometry, path, layer=0): # TODO: manually tested but no tests exist yet
+    """Read nightlight data from single NASA BlackMarble tile
+    and crop to given shape.
+
+    Parameters
+    ----------
+    geometry : shape(s) to crop data to in degree lon/lat.
+        for example shapely.geometry.Polygon(s) object or
+        from polygon defined in a shapefile.
+    path : Path or str
+        full path to BlackMarble tif (including filename)
+    layer : int
+        TIF-layer to be returned. The default is 0.
+        BlackMarble usually comes with 3 layers.
+
+    Returns
+    -------
+    out_image[layer,:,:] : 2D numpy ndarray with cropped data
+    meta : dict containing meta data 
     """
-    # TODO: argument req_files is not used in this function
-
-    coord_min = np.array([-90, -180]) + NASA_RESOLUTION_DEG / 2
-    coord_h = np.full((2,), NASA_RESOLUTION_DEG)
-
-    min_lon, min_lat, max_lon, max_lat = bounds
-    bounds_mat = np.array([[min_lat, min_lon], [max_lat, max_lon]])
-    global_idx = (bounds_mat - coord_min[None]) / coord_h[None]
-    global_idx[0, :] = np.floor(global_idx[0, :])
-    global_idx[1, :] = np.ceil(global_idx[1, :])
-    tile_size = np.array(NASA_TILE_SIZE)
-
-    nightlight = []
-    for idx, fname in enumerate(BM_FILENAMES):
-        tile_coord = np.array([1 - idx % 2, idx // 2])
-        extent = global_idx - (tile_coord * tile_size)[None]
-        if np.any(extent[1, :] < 0) or np.any(extent[0, :] >= NASA_TILE_SIZE):
-            # this tile does not intersect the specified bounds
-            continue
-        extent = np.int64(np.clip(extent, 0, tile_size[None] - 1))
-
-        im_nl, _ = read_bm_file(SYSTEM_DIR, fname.replace('*', str(year)))
-        im_nl = np.flipud(im_nl)
-        im_nl = sparse.csc.csc_matrix(im_nl)
-        im_nl = im_nl[extent[0, 0]:extent[1, 0] + 1, extent[0, 1]:extent[1, 1] + 1]
-        nightlight.append((tile_coord, im_nl))
-
-    tile_coords = np.array([n[0] for n in nightlight])
-    shape = tile_coords.max(axis=0) - tile_coords.min(axis=0) + 1
-    nightlight = np.array([n[1] for n in nightlight]).reshape(shape, order='F')
-    nightlight = sparse.bmat(np.flipud(nightlight), format='csr')
-
-    coord_nl = np.vstack([coord_min, coord_h]).T
-    coord_nl[:, 0] += global_idx[0, :] * coord_h[:]
+    # open tif source file with raterio:
+    src = rasterio.open(path)
+    # read cropped data from  source file (src) to np.ndarray:
+    out_image, transform = rasterio.mask.mask(src, [geometry], crop=True)
+    meta = src.meta
+    meta.update({"driver": "GTiff",
+                 "height": out_image.shape[1],
+                 "width": out_image.shape[2],
+                 "transform": transform})
+    src.close()
+    return out_image[layer,:,:], meta
+
+# TODO: LitPop 2.0
+def load_nasa_nl_shape(geometry, reference_year, data_dir=None, dtype=None):
```

Comment:
I am confused by what this method is doing. Could you please write bullet point style what is (supposed to be) happening?

Reply:
added comments in docstr with https://github.com/CLIMADA-project/climada_python/pull/239/commits/d2ac04440a257c240723effb23996af72260f5f4